woocommerce/woocommerce-ios
# [Mobile Payments] Parse PaymentGateway instructions

Closes: #7479


Merge after #7519 

### Description


This PR updates the PaymentGateway model, and it's Core Data storage, to include the `instructions` property/attribute.

We are adding the option for merchants to enable the `Cash on Delivery` payment method by tapping a button on a new screen in In-Person Payments onboarding.

This button will update the PaymentGateway for the merchant's store, enabling it, and changing the `title`, `description`, and `instructions`. The first three of these were covered in #7519, this PR addresses the last one, `instructions`.

`instructions` are shown on the Thank You page after checkout, and in the customer-facing Confirmation email.

The new request is not used anywhere yet, so cannot be directly tested, however unit tests are provided.


### Testing instructions


Run unit tests.

Further testing isn't really required as it will be covered when the button is implemented, but if you want to see the request working, add the following to `InPersonPaymentsCodPaymentGatewayNotSetUpViewModel`:

```
func enableCod() {
            guard let siteID = siteID else {
                return completion()
            }
        let action = PaymentGatewayAction.updatePaymentGateway(PaymentGateway(siteID: siteID, gatewayID: "cod", title: "Pay in Person", description: "Pay in person by card or cash", enabled: true, features: [.products], instructions: "Pay in Person: you can use card or cash")) { result in
            guard let response = try? result.get() else {
                DDLogError("💰 Could not update Payment Gateway : \(String(describing: result.failure))")
                return
            }
            DDLogInfo("💰 Updated Payment Gateway : \(response)")
        }
        stores.dispatch(action)
    }
```
and then update the `InPersonPaymentsCodPaymentGatewayNotSetUp` by setting the button's action to
```
Button(Localization.enableButton, action: viewModel.enableCod)
                .buttonStyle(PrimaryLoadingButtonStyle(isLoading: false))
```

Then you can tap the Enable button on the new screen, and it will send the network request for you to inspect in the console, and/or a network proxy tool such as Proxyman.

Additionally, using a breakpoint in the `updatePaymentGateways` method at app launch, you can observe that `storageGateway.instructions` is nil the first time around, and matches the instruction `"Pay in Person: you can use card or cash"` as per the code above on second and subsequent app launches.

### Testing output

Here is the request and response from my testing

#### Request
```
POST /rest/v1.1/jetpack-blogs/200370280/rest-api/ HTTP/1.1
Host: public-api.wordpress.com
Content-Type: application/x-www-form-urlencoded; charset=utf-8
Content-Length: 385
Connection: keep-alive
Accept: application/json
User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 15_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 wc-ios/9.9
Accept-Encoding: gzip;q=1.0, compress;q=0.5
Accept-Language: en;q=1.0

body=%7B%22method_supports%22%3A%5B%22products%22%5D%2C%22settings%22%3A%7B%22instructions%22%3A%22Pay%20in%20Person%3A%20you%20can%20pay%20by%20card%20or%20cash%22%7D%2C%22enabled%22%3Atrue%2C%22title%22%3A%22Pay%20in%20Person%22%2C%22description%22%3A%22Pay%20in%20person%20by%20card%20or%20cash%22%2C%22id%22%3A%22cod%22%7D&json=true&path=/wc/v3/payment_gateways/cod%26_method%3Dput
```

#### Response
```
HTTP/1.1 200 OK
Server: nginx
Date: Fri, 19 Aug 2022 11:59:37 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
Vary: Accept-Encoding
Host-Header: WordPress.com
X-hacker: Oh, Awesome: Opossum
Expires: Wed, 11 Jan 1984 05:00:00 GMT
Cache-Control: no-cache, must-revalidate, max-age=0
X-Jetpack-Proxy-Timing: 419.1210269928
Content-Encoding: gzip
X-ac: 2.lhr _dca 
Strict-Transport-Security: max-age=15552000

{
  "data": {
    "id": "cod",
    "title": "Pay in Person",
    "description": "Pay in person by card or cash",
    "order": 3,
    "enabled": true,
    "method_title": "Cash on delivery",
    "method_description": "Have your customers pay with cash (or by other means) upon delivery.",
    "method_supports": [
      "products"
    ],
    "settings": {
      "title": {
        "id": "title",
        "label": "Title",
        "description": "Payment method description that the customer will see on your checkout.",
        "type": "safe_text",
        "value": "Pay in Person",
        "default": "Cash on delivery",
        "tip": "Payment method description that the customer will see on your checkout.",
        "placeholder": ""
      },
      "instructions": {
        "id": "instructions",
        "label": "Instructions",
        "description": "Instructions that will be added to the thank you page.",
        "type": "textarea",
        "value": "Pay in Person: you can pay by card or cash",
        "default": "Pay with cash upon delivery.",
        "tip": "Instructions that will be added to the thank you page.",
        "placeholder": ""
      },
      "enable_for_methods": {
        "id": "enable_for_methods",
        "label": "Enable for shipping methods",
        "description": "If COD is only available for certain methods, set it up here. Leave blank to enable for all methods.",
        "type": "multiselect",
        "value": "",
        "default": "",
        "tip": "If COD is only available for certain methods, set it up here. Leave blank to enable for all methods.",
        "placeholder": "",
        "options": {
          "Flat rate": {
            "flat_rate": "Any &quot;Flat rate&quot; method",
            "flat_rate:4": "Iowa &ndash; Flat rate (#4)",
            "flat_rate:1": "United States (US) &ndash; Flat rate (#1)",
            "flat_rate:2": "Other locations &ndash; Flat rate (#2)"
          },
          "Free shipping": {
            "free_shipping": "Any &quot;Free shipping&quot; method"
          },
          "Local pickup": {
            "local_pickup": "Any &quot;Local pickup&quot; method",
            "local_pickup:5": "Iowa &ndash; Local pickup (#5)"
          }
        }
      },
      "enable_for_virtual": {
        "id": "enable_for_virtual",
        "label": "Accept COD if the order is virtual",
        "description": "",
        "type": "checkbox",
        "value": "yes",
        "default": "yes",
        "tip": "",
        "placeholder": ""
      }
    },
    "needs_setup": false,
    "post_install_scripts": [],
    "settings_url": "https://pointlessdivision.wpcomstaging.com/wp-admin/admin.php?page=wc-settings&tab=checkout&section=cod",
    "connection_url": null,
    "setup_help_text": null,
    "required_settings_keys": [],
    "_links": {
      "self": [
        {
          "href": "https://pointlessdivision.wpcomstaging.com/wp-json/wc/v3/payment_gateways/cod"
        }
      ],
      "collection": [
        {
          "href": "https://pointlessdivision.wpcomstaging.com/wp-json/wc/v3/payment_gateways"
        }
      ]
    }
  }
}
```

---
